### ISSUE TITLE
Cxx and CxxStdlib built for iOS 16.0 and above only

### ISSUE THREAD

**Issue by drodriguez** [swiftlang/swift#77909]
### Description

Cxx.swiftmodule and CxxStdlib.swiftmodule in recent nightlies of Swift.org and Xcode 16.2 betas seems to only be built for iOS 16.0 and above.

### Reproduction

With Xcode 16.2 beta 3:

1. Download Apple example code in https://developer.apple.com/documentation/swift/mixinglanguagesinanxcodeproject
2. Open in Xcode 16.2 beta 3 (install iOS SDK).
3. In the project settings, in Signing and Certificates, choose a team.
4.  In the project settings, in Build Settings, change the "Base SDK" to iOS.
5.  Same place, change the "iOS deployment target" to iOS 15 (minimum allowed by the UI)
6. Build the target "Swift-and-Cplusplus-mixed-FibonacciFramework"

Downloading Swift.org nightlies seems to reproduce the problem as well, but only since 2024-10-08 nightlies in the swift-6.0 branch.

Good: swift-6.0-DEVELOPMENT-SNAPSHOT-2024-09-17-a
Bad: swift-6.0-DEVELOPMENT-SNAPSHOT-2024-10-08-a
Also bad: swift-6.0-DEVELOPMENT-SNAPSHOT-2024-11-05-a (last available)

### Expected behavior

In Xcode 16.1 and Xcode 16.0 the previous code sample compiled without any problems.

### Environment

Reproducible with several toolchains:

- Xcode 16.2 beta 3
- swift-6.0-DEVELOPMENT-SNAPSHOT-2024-10-08-a
- swift-6.0-DEVELOPMENT-SNAPSHOT-2024-11-05-a

### Additional information

I think the commit that introduced the problem might be https://github.com/swiftlang/swift/commit/b87b263742d55a2187c188f5d3db1517ea5d3740, or at least the timeline of nightlies seems to points to it, and the change seems to match the symptoms. In Xcode 16.1, the `Cxx.swiftmodule/arm64e-apple-ios.swiftmodule` file has the following entry (dump from `llvm-bcanalyzer`):

```
<TARGET abbrevid=6/> blob data = 'arm64e-apple-ios11.0'
```

While the same file in Xcode 16.2b3 has the following:

```
<TARGET abbrevid=6/> blob data = 'arm64e-apple-ios16.0'
```

I think that Cxx.swiftmodule seems to "underspecify" its deployment version for anything but macOS in https://github.com/swiftlang/swift/blob/main/stdlib/public/Cxx/CMakeLists.txt#L28 ends up using the default `SWIFT_DARWIN_DEPLOYMENT_VERSION_IOS`.

It also might be related to Cxx not being resilient or not distributing `.swiftinterface` files.

I wrote asking about this in the Slack #cxx-interop channel. I will try to hang out there to answer questions.

We opened FB16023505 as well.

**Comment by egorzhdan** [comment]
Thanks @drodriguez for bringing this up!

**Comment by mrousavy** [comment]
Hey - I'm getting the same error now after upgrading to Xcode 16.2.:

```diff
- Compiling for iOS 15.1, but module 'CxxStdlib' has a minimum deployment target of iOS 16.0: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator/CxxStdlib.swiftmodule/arm64-apple-ios-simulator.swiftmodule
```

Does 16.2. work for you now? I'm not sure if the fix by @egorzhdan landed in the Xcode 16.2. release...

**Comment by drodriguez** [comment]
The final Xcode 16.2 release included a fix for the devices, but not for the simulators. You can find the following in the [release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-16_2-release-notes):

> Fixed: Apps using Swift/C++ interoperability may not be back deployable to operating systems earlier than iOS 16, watchOS 9, and tvOS 16. (140823785)

> Known Issue: Xcode targets using Swift/C++ interoperability and having a deployment target of less than iOS 16 or tvOS 16 may not build for iOS or tvOS simulators. (141232269)

**Comment by mrousavy** [comment]
Got it - thank you!

**Comment by mrousavy** [comment]
Got it - thank you!

**Comment by egorzhdan** [comment]
The Simulator issue should be fixed by https://github.com/swiftlang/swift/pull/78242

**Comment by RafeeqShodeinde** [comment]
> The Simulator issue should be fixed by #78242

Hello, please can you share the process for applying this fix you mentioned to Xcode 16.2? Where I can locate the "stdlib/cmake/modules/SwiftSource.cmake" file to add the changes for the ios simulator? Thanks in advance

**Comment by egorzhdan** [comment]
@RafeeqShodeinde this isn't something you could fix within an Xcode installation unfortunately. The pull request that I mentioned above changes the build scripts that are used to build Swift toolchains. That means a toolchain needs to be rebuilt with this change applied for the fix to take effect.

**Comment by ErwannRousseau** [comment]
Hi @mrousavy I'm trying [react-native-nitro-modules](https://github.com/mrousavy/nitro/blob/main/packages/react-native-nitro-modules) and I have the same problem as you, but I'm not sure how to do what I need to do to solve this problem....
 
I don't know what you did to solve this problem, I'm totally stuck on this point even reading this issue 😕.
I hope you can help me, I wanted to open an issue on Nitro Module, but I understand that it does not come from your library

**Comment by nemanjamitric** [comment]
> Hi [@mrousavy](https://github.com/mrousavy) I'm trying [react-native-nitro-modules](https://github.com/mrousavy/nitro/blob/main/packages/react-native-nitro-modules) and I have the same problem as you, but I'm not sure how to do what I need to do to solve this problem....
> 
> I don't know what you did to solve this problem, I'm totally stuck on this point even reading this issue 😕. I hope you can help me, I wanted to open an issue on Nitro Module, but I understand that it does not come from your library

Did you find a solution?

**Comment by ErwannRousseau** [comment]
@nemanjamitric Hi !

sorry for the delay I was working on other things 

I've just tried again, and I'm still struggling with the same errer 

```
compiling for iOS 15.1, but module 'CxxStdlib' has a minimum deployment target of iOS 16.0: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator/CxxStdlib.swiftmodule/x86_64-apple-ios-simulator.swiftmodule
```

**Comment by nemanjamitric** [comment]
> [@nemanjamitric](https://github.com/nemanjamitric) Hi !
> 
> sorry for the delay I was working on other things
> 
> I've just tried again, and I'm still struggling with the same errer
> 
> ```
> compiling for iOS 15.1, but module 'CxxStdlib' has a minimum deployment target of iOS 16.0: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator/CxxStdlib.swiftmodule/x86_64-apple-ios-simulator.swiftmodule
> ```

I've fixed it since. What you need to do is add a config file:
```
const { withDangerousMod } = require('@expo/config-plugins');
const {
  mergeContents,
} = require('@expo/config-plugins/build/utils/generateCode');
const fs = require('fs');
const path = require('path');

const withIosDeploymentTarget = (config) => {
  return withDangerousMod(config, [
    'ios',
    async (config) => {
      // Find the Podfile
      const podfile = path.join(
        config.modRequest.platformProjectRoot,
        'Podfile',
      );
      // Read the Podfile
      const podfileContents = fs.readFileSync(podfile, 'utf8');
      // Merge the contents of the Podfile with the new content setting
      // the deployment target of all targets to 16.0
      const setDeploymentTarget = mergeContents({
        tag: 'ios-deployment-target',
        src: podfileContents,
        newSrc: `    installer.pods_project.targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = '16.0'
      end
    end`,
        anchor: /post_install do \|installer\|/i,
        offset: 1,
        comment: '#',
      });

      if (!setDeploymentTarget.didMerge) {
        return config;
      }

      fs.writeFileSync(podfile, setDeploymentTarget.contents);

      return config;
    },
  ]);
};

module.exports = withIosDeploymentTarget;
```
and add it to app.json
```
{
    "expo": {
        ...,
        "plugins": [
            ...,
            "./path/to/configFile.js"
        ]
        ...
    }
}
```

**Important: I didn't write this script, I can't find where I got it from, I would love to credit whoever first wrote it.**

**Comment by kickbk** [comment]
@ErwannRousseau don't give up just yet. I shared an updated config plugin [here](https://github.com/mrousavy/nitro/issues/422#issuecomment-2682880817) which sets the minimum deployment target at build. Then running `npx expo prebuild` and `npx expo run:ios` builds and launches the app on the simulator (I only tested iOS so far). I'm using unistyles v3. If you get any unistyles errors, those will need to be addressed, but I can confirm I got it working.
